--- Makefile	2008-07-14 05:32:14.000000000 -0500
+++ Makefile.new	2008-07-14 05:32:08.000000000 -0500
@@ -52,11 +52,12 @@
 
 # LDFLAGS sets ${PREFIX}/lib as the LD_LIBRARY in the binary
 # CB: rpath is ignored on OS X (so where are the library files taken from?). 
-python: ${PYTHON}.tgz tk blt
+python: ${PYTHON}.tgz
 	echo ${PREFIX}
 	tar -xzf ${PYTHON}.tgz
 # Configure for specific operating system versions
 ifeq ("$(shell uname -s)","Darwin")
+	patch -p0 < Python_OSX_tk.diff
 ifeq ("$(shell uname -r | cut -d . -f 1)","9") # Darwin 9.x == Mac OS X 10.5 (Leopard)
 	cd ${PYTHON}; ./configure --prefix=${PREFIX} MACOSX_DEPLOYMENT_TARGET=10.5
 else
