Index: Makefile
===================================================================
--- Makefile	(revision 7767)
+++ Makefile	(working copy)
@@ -29,6 +29,7 @@
 python: ${PYTHON}.tgz tk
 	echo ${PREFIX}
 	tar -xozf ${PYTHON}.tgz
+	patch -p0 < Python_OSX_X11.diff
 	cd ${PYTHON}; ./configure --prefix=${PREFIX} LDFLAGS=-Wl,-rpath,${PREFIX}/lib
 	cd ${PYTHON}; env PREFIX=${PREFIX} make 
 	cd ${PYTHON}; env PREFIX=${PREFIX} make install
@@ -141,6 +142,7 @@
 ifeq ("$(shell uname -m)","x86_64")
 	cd ${PIL}; patch < ../Imaging.diff
 endif
+	patch -p0 < Imaging_OSX.diff
 	cd ${PIL}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${PREFIX}/bin/python setup.py build_ext -i
 	cd ${PIL}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${PREFIX}/bin/python selftest.py
 	cd ${PIL}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${PREFIX}/bin/python setup.py install
