*** Makefile	Wed Jul 23 15:22:31 2008
--- Makefile.win	Wed Jul 23 15:22:26 2008
***************
*** 1,13 ****
! PREFIX = ${dir ${CURDIR}}
  CC = gcc
  
  # Options to pass to all setup.py install commands.
  # (Allows us to control installation location even when not using our
  # own Python, e.g. on Windows.)
! PYSETUP_OPT = --prefix=${PREFIX}
! 
! default: python jpeg pil numpy matplotlib weave fixedpoint pychecker common pylint epydoc docutils gnosis ipython gmpy tilewrapper scrodget tooltip
  
  
  all: default pyaudiolab mlabwrap scipy ipython pyrex pytables psyco gnuplotpy
  
--- 1,16 ----
! PREFIX = $(shell echo ${dir ${CURDIR}} | sed 's/\///'  | sed 's/\//:\//' | sed 's/\//\\\\/g' | sed 's/\(.*\)./\1/')
! PREFIX_FWD = $(shell echo ${dir ${CURDIR}} | sed 's/\///'  | sed 's/\//:\//' | sed 's/\(.*\)./\1/')
  CC = gcc
+ # PIL
+ # http://mail.python.org/pipermail/image-sig/2004-December/003067.html
  
  # Options to pass to all setup.py install commands.
  # (Allows us to control installation location even when not using our
  # own Python, e.g. on Windows.)
! PYSETUP_OPT = --prefix=${PREFIX} --install-scripts=${PREFIX}\\bin
  
+ default: numpy matplotlib weave fixedpoint pychecker common pylint epydoc docutils gnosis ipython gmpy tilewrapper scrodget tooltip
+ # i.e. python, jpeg, pil installed as binaries beforehand
  
  all: default pyaudiolab mlabwrap scipy ipython pyrex pytables psyco gnuplotpy
  
***************
*** 18,24 ****
  uninstall: python-uninstall jpeg-uninstall pil-uninstall numpy-uninstall tk-uninstall tcl-uninstall matplotlib-uninstall fixedpoint-uninstall weave-uninstall pychecker-uninstall common-uninstall pylint-uninstall epydoc-uninstall docutils-uninstall gnosis-uninstall scipy-uninstall mlabwrap-uninstall ipython-uninstall libsndfile-uninstall pyaudiolab-uninstall tilewrapper-uninstall psyco-uninstall pyrex-uninstall hdf5-uninstall pytables-uninstall gnuplotpy-uninstall player-uninstall progressbar-uninstall stage-uninstall gmp-uninstall mxnumber-uninstall gmpy-uninstall scrodget-uninstall tooltip-uninstall snit-uninstall
  
  
- 
  ##############################################################
  # libblt: already available on nearly all systems, but some systems
  # have a version incompatible with tkinter (and Tcl/Tk?), so we build
--- 21,26 ----
***************
*** 188,194 ****
  #       # Tell OS X machines where to find freetype
  	patch -p0 < Imaging_OSX.diff
  endif
! 	cd ${PIL}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${PREFIX}/bin/python setup.py build_ext -i
  	cd ${PIL}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${PREFIX}/bin/python selftest.py
  	cd ${PIL}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${PREFIX}/bin/python setup.py install ${PYSETUP_OPT}
  	touch pil
--- 190,196 ----
  #       # Tell OS X machines where to find freetype
  	patch -p0 < Imaging_OSX.diff
  endif
! 	cd ${PIL}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${PREFIX}/bin/python setup.py build_ext --compiler=mingw32 -i
  	cd ${PIL}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${PREFIX}/bin/python selftest.py
  	cd ${PIL}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${PREFIX}/bin/python setup.py install ${PYSETUP_OPT}
  	touch pil
***************
*** 236,242 ****
  
  matplotlib: ${MATPLOTLIB}.tar.gz python
  	tar -xzf ${MATPLOTLIB}.tar.gz
! 	cd ${MATPLOTLIB}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${MATPLOTLIB_DISPLAY} ${PREFIX}/bin/python setup.py build
  	cd ${MATPLOTLIB}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${MATPLOTLIB_DISPLAY} ${PREFIX}/bin/python setup.py install ${PYSETUP_OPT}
  	touch matplotlib
  
--- 238,246 ----
  
  matplotlib: ${MATPLOTLIB}.tar.gz python
  	tar -xzf ${MATPLOTLIB}.tar.gz
! 	cp win32_static.tar.gz ${MATPLOTLIB}
! 	cd ${MATPLOTLIB}; tar -xzf win32_static.tar.gz
! 	cd ${MATPLOTLIB}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${MATPLOTLIB_DISPLAY} ${PREFIX}/bin/python setup.py build --compiler=mingw32 bdist_wininst
  	cd ${MATPLOTLIB}; env PREFIX=${PREFIX} LD_LIBRARY_PATH=${PREFIX}/lib ${MATPLOTLIB_DISPLAY} ${PREFIX}/bin/python setup.py install ${PYSETUP_OPT}
  	touch matplotlib
  
***************
*** 509,515 ****
  TILEWRAPPER = Tile
  
  tilewrapper: python
! 	cp Tile.py ${PREFIX}/lib/python2.5/site-packages/
  	touch tilewrapper
  
  tilewrapper-uninstall:
--- 513,519 ----
  TILEWRAPPER = Tile
  
  tilewrapper: python
! 	cp Tile.py ${PREFIX}/Lib/site-packages
  	touch tilewrapper
  
  tilewrapper-uninstall:
***************
*** 760,773 ****
  GMP_UNINSTALL += lib/libgmp.* 
  GMP_UNINSTALL += include/gmp.h
  
- 
  gmp: ${GMP}.tar.gz
  	tar -xzf ${GMP}.tar.gz
! 	cd ${GMP}; ./configure --prefix=${PREFIX} 
  	make -C ${GMP}
  	make -C ${GMP} check
  	make -C ${GMP} install
  	touch gmp
  
  gmp-clean: 
  	rm -rf ${GMP} 
--- 764,777 ----
  GMP_UNINSTALL += lib/libgmp.* 
  GMP_UNINSTALL += include/gmp.h
  
  gmp: ${GMP}.tar.gz
  	tar -xzf ${GMP}.tar.gz
! 	cd ${GMP}; ./configure --prefix=${PREFIX_FWD} 
  	make -C ${GMP}
  	make -C ${GMP} check
  	make -C ${GMP} install
  	touch gmp
+ # need to reset PREFIX!
  
  gmp-clean: 
  	rm -rf ${GMP} 
***************
*** 808,814 ****
  
  gmpy: ${GMPY}.tar.gz python gmp
  	tar -xzf ${GMPY}.tar.gz
! 	cd ${GMPY}; ${PREFIX}/bin/python setup.py build_ext --include-dirs='${PREFIX}/include' --library-dirs='${PREFIX}/lib'; ${PREFIX}/bin/python setup.py install ${PYSETUP_OPT}
  	touch gmpy
  
  gmpy-clean:
--- 812,818 ----
  
  gmpy: ${GMPY}.tar.gz python gmp
  	tar -xzf ${GMPY}.tar.gz
! 	cd ${GMPY}; ${PREFIX}/bin/python setup.py build_ext --compiler=mingw32 --include-dirs='${PREFIX}/include' --library-dirs='${PREFIX}/lib'; ${PREFIX}/bin/python setup.py install ${PYSETUP_OPT}
  	touch gmpy
  
  gmpy-clean:
