--- numpy/testing/numpytest.py	2007-04-12 18:53:46.000000000 +0100
+++ numpy/testing/numpytest.py	2007-03-05 18:48:14.000000000 +0000
@@ -194,8 +194,8 @@
 
 class ScipyTestCase(NumpyTestCase):
     def __init__(self, package=None):
-        warnings.warn("ScipyTestCase is now called NumpyTestCase; please update your code",
-                         DeprecationWarning)
+#        warnings.warn("ScipyTestCase is now called NumpyTestCase; please update your code",
+#                         DeprecationWarning)
         NumpyTestCase.__init__(self, package)
 
 
@@ -630,8 +630,8 @@
 
 class ScipyTest(NumpyTest):
     def __init__(self, package=None):
-        warnings.warn("ScipyTest is now called NumpyTest; please update your code",
-                         DeprecationWarning)
+#        warnings.warn("ScipyTest is now called NumpyTest; please update your code",
+#                         DeprecationWarning)
         NumpyTest.__init__(self, package)
 
 
