--- numpy/testing/numpytest.py.new	2008-08-18 17:13:25.702227000 +0100
+++ numpy/testing/numpytest.py	2008-08-18 17:12:34.319247000 +0100
@@ -198,8 +198,8 @@ class NumpyTestCase (unittest.TestCase):
 
 class ScipyTestCase(NumpyTestCase):
     def __init__(self, package=None):
-        warnings.warn("ScipyTestCase is now called NumpyTestCase; please update your code",
-                         DeprecationWarning, stacklevel=2)
+        #warnings.warn("ScipyTestCase is now called NumpyTestCase; please update your code",
+        #                 DeprecationWarning, stacklevel=2)
         NumpyTestCase.__init__(self, package)
 
 
@@ -668,8 +668,8 @@ class NumpyTest:
 
 class ScipyTest(NumpyTest):
     def __init__(self, package=None):
-        warnings.warn("ScipyTest is now called NumpyTest; please update your code",
-                         DeprecationWarning, stacklevel=2)
+        #warnings.warn("ScipyTest is now called NumpyTest; please update your code",
+        #                 DeprecationWarning, stacklevel=2)
         NumpyTest.__init__(self, package)
 
 
