diff -u Pmw_1_2/lib/PmwBase.py~ Pmw_1_2/lib/PmwBase.py  
--- Pmw_1_2/lib/PmwBase.py~	2003-05-17 15:14:08.000000000 +0100
+++ Pmw_1_2/lib/PmwBase.py	2007-12-12 16:21:02.000000000 +0000
@@ -1771,8 +1771,8 @@
     exc_type, exc_value, exc_traceback = sys.exc_info()
 
     # Give basic information about the callback exception.
-    if type(exc_type) == types.ClassType:
+    if type(exc_type) == types.ClassType or isinstance(exc_type,type):
 	# Handle python 1.5 class exceptions.
 	exc_type = exc_type.__name__
     msg = exc_type + ' Exception in Tk callback\n'
 
